source/Plugins/Language/Swift/SwiftCodeCompletion.cpp

```diff
@@ -192,8 +192,166 @@ doCodeCompletion(SourceFile &SF, StringRef EnteredCode,
+    if (!CompletionsModule)
+      return CompletionResponse::error("could not make completions module");
+
+    // This file accumulates all of the "hand imports" (imports that the user
+    // made in previous executions). We also put the current entered code in
+    // this file.
+    AddSourceFile(Ctx, CompletionsModule, SourceFileKind::REPL);
+
+    // We reset this file with the persistent decls every completion request.
+    AddSourceFile(Ctx, CompletionsModule, SourceFileKind::Library);
+  }
+
+  // This file accumulates all of the "hand imports" (imports that the user
+  // made in previous executions). We also put the current entered code in
+  // this file.
+  SourceFile *EnteredCodeFile =
+      GetSingleSourceFile(CompletionsModule, SourceFileKind::REPL);
+  assert(EnteredCodeFile);
+
+  // Accumulate new hand imports into the file.
+  {
+    // First, construct a set of imports that we already have, so that we can
+    // avoid duplicate imports.
+    std::set<ModuleDecl *> ExistingImportSet;
+    SmallVector<ModuleDecl::ImportedModule, 8> ExistingImports;
+    EnteredCodeFile->getImportedModules(ExistingImports,
+                                        ModuleDecl::ImportFilter::All);
+    for (auto &ExistingImport : ExistingImports)
+      ExistingImportSet.insert(std::get<1>(ExistingImport));
+
+    // Next, add new imports into the file.
+    SmallVector<SourceFile::ImportedModuleDesc, 8> NewImports;
+    PersistentExpressionState.RunOverHandLoadedModules(
+        [&](const ConstString ModuleName) -> bool {
+          // Skip these modules, to prevent "TestSwiftCompletions.py" from
+          // segfaulting.
+          // TODO: Investigate why this happens.
+          if (ModuleName == ConstString("SwiftOnoneSupport"))
+            return true;
+          if (ModuleName == ConstString("a"))
+            return true;
+
+          ModuleDecl *Module = SwiftCtx.GetModule(ModuleName, Error);
+          if (!Module)
+            return true;
+          if (ExistingImportSet.find(Module) != ExistingImportSet.end())
+            return true;
+          NewImports.push_back(SourceFile::ImportedModuleDesc(
+              std::make_pair(ModuleDecl::AccessPathTy(), Module),
+              SourceFile::ImportOptions()));
+        });
+    EnteredCodeFile->addImports(NewImports);
+  }
+
+  // We reset this file with the persistent decls every completion request.
+  SourceFile *PreviousDeclsFile =
+      GetSingleSourceFile(CompletionsModule, SourceFileKind::Library);
+  assert(PreviousDeclsFile);
+
+  // Reset the decls to the persistent decls.
+  {
+    std::vector<Decl *> PersistentDecls;
+    PersistentExpressionState.GetAllDecls(PersistentDecls);
+    PreviousDeclsFile->Decls.clear();
+    for (auto *PersistentDecl : PersistentDecls)
+      PreviousDeclsFile->Decls.push_back(PersistentDecl);
+    PreviousDeclsFile->clearLookupCache();
+  }
+
+  // `PreviousDeclsFile` might now contain decls that a re-defined in
```

Comment:
This approach might not work correctly with extensions. e.g., if you redefine a nested type in an extension, you might still have both the definitions in the auto complete. 

I think a proper fix for https://bugs.swift.org/browse/TF-10 shoud take care of the issue automatically.

Reply:
Indeed, it behaves weirdly with extensions. I have added some tests documenting the current behavior. We'll defer actually fixing it until later.